homeassistant/components/bluetooth_tracker/device_tracker.py

```diff
@@ -65,99 +72,107 @@ def discover_devices(device_id: int) -> List[Tuple[str, str]]:
+        device
+        for device in device_registry_instance.devices.values()
+        for domain, mac in device.identifiers
+        if domain == DOMAIN
+    )
 
-    # If track new devices is true discover new devices on startup.
-    track_new: bool = config.get(CONF_TRACK_NEW, DEFAULT_TRACK_NEW)
-    _LOGGER.debug("Tracking new devices is set to %s", track_new)
+    entities: Set[BluetoothEntity] = set()
+    for device in devices:
+        hass.data[DOMAIN]["devices"].add(device.mac)
+        entity = BluetoothEntity(config_entry, device.mac, device.name, {})
+        entities.add(entity)
+
+    async_add_entities(list(entities))
+
+    return entities
 
-    devices_to_track, devices_to_not_track = await get_tracking_devices(hass)
 
-    if not devices_to_track and not track_new:
-        _LOGGER.debug("No Bluetooth devices to track and not tracking new devices")
+async def async_setup_entry(
+    hass: HomeAssistantType,
+    config_entry: ConfigEntry,
+    async_add_entities: Callable[List[Any]],
+):
+    """Set up a bluetooth device tracker."""
+    device_id: int = config_entry.data[CONF_DEVICE_ID]
+    interval = config_entry.options[CONF_SCAN_INTERVAL]
+    request_rssi: bool = config_entry.options[CONF_REQUEST_RSSI]
+    track_new: bool = config_entry.options[CONF_TRACK_NEW]
+    _LOGGER.debug("Tracking new devices is set to %s", track_new)
+
+    update_bluetooth_lock = asyncio.Lock()
+    entities = load_initial_entities(hass, config_entry, async_add_entities)
 
     if request_rssi:
         _LOGGER.debug("Detecting RSSI for devices")
 
+    def see_device(mac: str, device_name: str, rssi=None) -> None:
+        """Mark a device as seen."""
+        attributes = {}
+        if rssi is not None:
+            attributes["rssi"] = rssi
+
+        device = BluetoothEntity(config_entry, mac, device_name, attributes)
+        if device in entities:
+            async_dispatcher_send(hass, TRACKER_UPDATE, mac, device_name, attributes)
+            return
+
+        entities.add(device)
+        async_add_entities([device])
+
     async def perform_bluetooth_update():
         """Discover Bluetooth devices and update status."""
-
         _LOGGER.debug("Performing Bluetooth devices discovery and update")
-        tasks = []
+
+        tracked_devices = set()
 
         try:
             if track_new:
                 devices = await hass.async_add_executor_job(discover_devices, device_id)
                 for mac, device_name in devices:
-                    if mac not in devices_to_track and mac not in devices_to_not_track:
-                        devices_to_track.add(mac)
-
-            for mac in devices_to_track:
-                device_name = await hass.async_add_executor_job(lookup_name, mac)
+                    device = next((d for d in entities if d.unique_id == mac), None)
+                    if not device:
+                        device = BluetoothEntity(config_entry, mac, device_name, {})
```

Comment:
We create the entities in `see_device`. We should decide on one place where we do this.